File: core/transport/connection_manager.go
package transport

import (
	"encoding/json"
	"fmt"
	"math/rand"
	"net"
	"sync"
	"time"

	"wedrop/core/discovery"
	"wedrop/core/protocol"
)

const (
	dialTimeout    = 4 * time.Second
	reconnectEvery = 3 * time.Second
	minBackoff     = 2 * time.Second
	maxBackoff     = 60 * time.Second
)

// PairingDecision is what the UI returns when a device asks to join.
type PairingDecision struct {
	Accepted bool
	Reason   string
}

// Manager owns the TCP listener, keeps a session open to every trusted peer
// that is online, and routes inbound connections by intent.
//
// The dialling loop deliberately never holds the session lock across a network
// call. The previous version dialled with the mutex held, so one unreachable
// peer froze clipboard sync and the device list for the full dial timeout on
// every tick.
type Manager struct {
	local     LocalInfo
	discovery *discovery.Service
	auth      PeerAuthorizer

	mu       sync.RWMutex
	sessions map[string]*sessionEntry
	dialing  map[string]struct{}
	backoff  map[string]*backoffState

	listener net.Listener
	port     int

	stopChan chan struct{}
	stopOnce sync.Once
	wg       sync.WaitGroup

	// Handler receives control messages from every session.
	Handler SessionHandler
	// OnPairingRequest is called when an untrusted device asks to join. It is
	// invoked on its own goroutine and may block while the user decides.
	OnPairingRequest func(req PairingRequest) PairingDecision
	// OnTransferOffer is called for an inbound transfer connection. The callee
	// owns the connection and must close it.
	OnTransferOffer func(conn *SecureConn, peer protocol.DeviceInfo, offer protocol.TransferOffer)
	// OnSessionChange fires whenever a peer connects or disconnects.
	OnSessionChange func(deviceID string, connected bool)
	// LocalDeviceInfo supplies the device info sent when a session opens.
	LocalDeviceInfo func() protocol.DeviceInfo
	// Logf receives diagnostics; nil disables logging.
	Logf func(format string, args ...interface{})
}

// PairingRequest describes an authenticated device asking to be added.
type PairingRequest struct {
	DeviceID         string
	Name             string
	Platform         string
	FormFactor       protocol.FormFactor
	PublicKey        string
	VerificationCode string
	Address          string
}

type sessionEntry struct {
	session  *Session
	outbound bool
}

type backoffState struct {
	next     time.Time
	duration time.Duration
}

// NewManager creates a connection manager. It does not touch the network until
// Start is called.
func NewManager(local LocalInfo, disc *discovery.Service, auth PeerAuthorizer) *Manager {
	return &Manager{
		local:     local,
		discovery: disc,
		auth:      auth,
		sessions:  make(map[string]*sessionEntry),
		dialing:   make(map[string]struct{}),
		backoff:   make(map[string]*backoffState),
		stopChan:  make(chan struct{}),
	}
}

// Start binds the listener and begins maintaining sessions. It returns the port
// actually bound, which the caller should announce over discovery — the old
// code advertised a hard-coded port even when the bind had failed, so peers
// spent their time dialling a socket that was not there.
func (m *Manager) Start() (int, error) {
	listener, err := net.Listen("tcp", fmt.Sprintf(":%d", protocol.TransportPort))
	if err != nil {
		// Fall back to an ephemeral port so a second instance, or a leftover
		// socket in TIME_WAIT, degrades instead of disabling the app.
		listener, err = net.Listen("tcp", ":0")
		if err != nil {
			return 0, fmt.Errorf("could not open a listening socket: %w", err)
		}
	}
	m.listener = listener
	m.port = listener.Addr().(*net.TCPAddr).Port

	m.wg.Add(2)
	go m.acceptLoop()
	go m.maintainLoop()
	return m.port, nil
}

// Port is the TCP port peers should connect to.
func (m *Manager) Port() int { return m.port }

// Stop closes the listener and every session.
func (m *Manager) Stop() {
	m.stopOnce.Do(func() {
		close(m.stopChan)
		if m.listener != nil {
			m.listener.Close()
		}

		m.mu.Lock()
		sessions := make([]*sessionEntry, 0, len(m.sessions))
		for _, e := range m.sessions {
			sessions = append(sessions, e)
		}
		m.sessions = make(map[string]*sessionEntry)
		m.mu.Unlock()

		for _, e := range sessions {
			e.session.Close()
		}
		m.wg.Wait()
	})
}

func (m *Manager) logf(format string, args ...interface{}) {
	if m.Logf != nil {
		m.Logf(format, args...)
	}
}

// ---------------------------------------------------------------- inbound

func (m *Manager) acceptLoop() {
	defer m.wg.Done()

	for {
		conn, err := m.listener.Accept()
		if err != nil {
			select {
			case <-m.stopChan:
				return
			default:
			}
			// A transient accept error should not kill the listener, but
			// spinning on a permanent one would burn a core.
			time.Sleep(100 * time.Millisecond)
			continue
		}
		go m.handleInbound(conn)
	}
}

func (m *Manager) handleInbound(conn net.Conn) {
	result, err := Accept(conn, m.local, m.auth)
	if err != nil {
		m.logf("rejected connection from %s: %v", conn.RemoteAddr(), err)
		conn.Close()
		return
	}

	switch result.Intent {
	case protocol.IntentPair:
		m.handlePairing(result, conn.RemoteAddr().String())
	case protocol.IntentSession:
		m.adoptSession(result, false)
	case protocol.IntentTransfer:
		m.handleInboundTransfer(result)
	default:
		result.Conn.Close()
	}
}

func (m *Manager) handlePairing(result *HandshakeResult, address string) {
	defer result.Conn.Close()

	if m.OnPairingRequest == nil {
		result.Conn.WriteJSON(protocol.PairingResp{
			Type: protocol.TypePairingResp, DeviceID: m.local.Identity.DeviceID,
			Accepted: false, Reason: "device cannot accept pairing right now",
		})
		return
	}

	decision := m.OnPairingRequest(PairingRequest{
		DeviceID:         result.Peer.DeviceID,
		Name:             result.Peer.Name,
		Platform:         result.Peer.Platform,
		FormFactor:       result.Peer.FormFactor,
		PublicKey:        result.PeerPublicKey,
		VerificationCode: result.VerificationCode,
		Address:          address,
	})

	local := protocol.DeviceInfo{}
	if m.LocalDeviceInfo != nil {
		local = m.LocalDeviceInfo()
	}

	result.Conn.WriteJSON(protocol.PairingResp{
		Type:     protocol.TypePairingResp,
		DeviceID: m.local.Identity.DeviceID,
		Name:     local.Name,
		Accepted: decision.Accepted,
		Reason:   decision.Reason,
	})

	if decision.Accepted {
		// Connect straight away rather than waiting for the next maintenance
		// tick, so the device turns green the moment the user taps Accept.
		m.AnnounceAndReconnect(result.Peer.DeviceID)
	}
}

func (m *Manager) handleInboundTransfer(result *HandshakeResult) {
	if m.OnTransferOffer == nil {
		result.Conn.Close()
		return
	}

	result.Conn.SetReadDeadline(time.Now().Add(30 * time.Second))
	msgBytes, err := result.Conn.ReadEncrypted()
	if err != nil {
		result.Conn.Close()
		return
	}

	msgType, err := protocol.ParseMessageType(msgBytes)
	if err != nil || msgType != protocol.TypeTransferOffer {
		result.Conn.Close()
		return
	}

	var offer protocol.TransferOffer
	if err := json.Unmarshal(msgBytes, &offer); err != nil {
		result.Conn.Close()
		return
	}

	// The callee owns the connection from here, including closing it.
	m.OnTransferOffer(result.Conn, result.Peer, offer)
}

// ---------------------------------------------------------------- outbound

func (m *Manager) maintainLoop() {
	defer m.wg.Done()

	ticker := time.NewTicker(reconnectEvery)
	defer ticker.Stop()

	for {
		select {
		case <-m.stopChan:
			return
		case <-ticker.C:
			m.connectMissing()
		}
	}
}

// connectMissing dials every trusted, online peer we are not already talking
// to. Each dial runs on its own goroutine so a slow peer never delays the rest.
func (m *Manager) connectMissing() {
	for _, peer := range m.discovery.Peers() {
		key, trusted := m.auth.TrustedKey(peer.DeviceID)
		if !trusted {
			continue
		}
		if !m.shouldDial(peer.DeviceID) {
			continue
		}

		m.mu.Lock()
		m.dialing[peer.DeviceID] = struct{}{}
		m.mu.Unlock()

		go m.dialPeer(peer, key)
	}
}

func (m *Manager) shouldDial(deviceID string) bool {
	m.mu.RLock()
	defer m.mu.RUnlock()

	if _, connected := m.sessions[deviceID]; connected {
		return false
	}
	if _, inFlight := m.dialing[deviceID]; inFlight {
		return false
	}
	if b, ok := m.backoff[deviceID]; ok && time.Now().Before(b.next) {
		return false
	}
	return true
}

func (m *Manager) dialPeer(peer discovery.Peer, expectedKey string) {
	defer func() {
		m.mu.Lock()
		delete(m.dialing, peer.DeviceID)
		m.mu.Unlock()
	}()

	address := net.JoinHostPort(peer.IP, fmt.Sprint(peer.TCPPort))
	result, err := Dial(address, m.local, protocol.IntentSession, expectedKey, dialTimeout)
	if err != nil {
		m.noteFailure(peer.DeviceID)
		m.logf("connect to %s (%s) failed: %v", peer.Name, address, err)
		return
	}

	m.clearBackoff(peer.DeviceID)
	m.adoptSession(result, true)
}

func (m *Manager) noteFailure(deviceID string) {
	m.mu.Lock()
	defer m.mu.Unlock()

	b, ok := m.backoff[deviceID]
	if !ok {
		b = &backoffState{duration: minBackoff}
		m.backoff[deviceID] = b
	} else {
		b.duration *= 2
		if b.duration > maxBackoff {
			b.duration = maxBackoff
		}
	}
	// Jitter keeps a roomful of devices from retrying in lockstep after the
	// Wi-Fi access point restarts.
	jitter := time.Duration(rand.Int63n(int64(b.duration / 4)))
	b.next = time.Now().Add(b.duration + jitter)
}

func (m *Manager) clearBackoff(deviceID string) {
	m.mu.Lock()
	delete(m.backoff, deviceID)
	m.mu.Unlock()
}

// AnnounceAndReconnect clears any backoff for a device and tries it at once.
func (m *Manager) AnnounceAndReconnect(deviceID string) {
	m.clearBackoff(deviceID)
	m.discovery.AnnounceNow()

	peer, ok := m.discovery.Peer(deviceID)
	if !ok {
		return
	}
	key, trusted := m.auth.TrustedKey(deviceID)
	if !trusted || !m.shouldDial(deviceID) {
		return
	}

	m.mu.Lock()
	m.dialing[deviceID] = struct{}{}
	m.mu.Unlock()

	go m.dialPeer(peer, key)
}

// ---------------------------------------------------------------- sessions

// adoptSession installs a new session, resolving the case where both devices
// dialled each other at the same moment.
//
// Both ends apply the same rule — keep the connection dialled by whichever
// device has the smaller ID — so they independently agree on which socket
// survives, instead of each closing the other's and ending up with none.
func (m *Manager) adoptSession(result *HandshakeResult, outbound bool) {
	deviceID := result.Peer.DeviceID

	m.mu.Lock()
	if existing, ok := m.sessions[deviceID]; ok {
		preferOutbound := m.local.Identity.DeviceID < deviceID
		if existing.outbound == preferOutbound || outbound != preferOutbound {
			// The session we already have is the preferred one (or this new
			// one is not); drop the duplicate.
			m.mu.Unlock()
			result.Conn.Close()
			return
		}
		delete(m.sessions, deviceID)
		m.mu.Unlock()
		existing.session.Close()
		m.mu.Lock()
	}

	session := NewSession(result.Conn, result.Peer, m)
	m.sessions[deviceID] = &sessionEntry{session: session, outbound: outbound}
	m.mu.Unlock()

	if m.LocalDeviceInfo != nil {
		session.Send(m.LocalDeviceInfo())
	}
	if m.OnSessionChange != nil {
		m.OnSessionChange(deviceID, true)
	}

	go session.Run()
}

// Session returns the live session for a device, if any.
func (m *Manager) Session(deviceID string) (*Session, bool) {
	m.mu.RLock()
	defer m.mu.RUnlock()

	e, ok := m.sessions[deviceID]
	if !ok {
		return nil, false
	}
	return e.session, true
}

// ConnectedDevices lists the device IDs with a live session.
func (m *Manager) ConnectedDevices() []string {
	m.mu.RLock()
	defer m.mu.RUnlock()

	out := make([]string, 0, len(m.sessions))
	for id := range m.sessions {
		out = append(out, id)
	}
	return out
}

// IsConnected reports whether a session is live for a device.
func (m *Manager) IsConnected(deviceID string) bool {
	m.mu.RLock()
	defer m.mu.RUnlock()
	_, ok := m.sessions[deviceID]
	return ok
}

// Disconnect closes the session with a device, if one is open.
func (m *Manager) Disconnect(deviceID string) {
	m.mu.Lock()
	e, ok := m.sessions[deviceID]
	delete(m.sessions, deviceID)
	m.mu.Unlock()

	if ok {
		e.session.Close()
	}
}

// Broadcast sends a message to every connected peer that advertised the given
// capability. An empty capability sends to everyone.
func (m *Manager) Broadcast(capability string, v interface{}) {
	m.mu.RLock()
	sessions := make([]*Session, 0, len(m.sessions))
	for _, e := range m.sessions {
		sessions = append(sessions, e.session)
	}
	m.mu.RUnlock()

	for _, s := range sessions {
		if capability != "" && !s.Supports(capability) {
			continue
		}
		if err := s.Send(v); err != nil {
			m.logf("send to %s failed: %v", s.DeviceID(), err)
			s.Close()
		}
	}
}

// SendTo delivers a message to one peer.
func (m *Manager) SendTo(deviceID string, v interface{}) error {
	session, ok := m.Session(deviceID)
	if !ok {
		return fmt.Errorf("device is not connected")
	}
	return session.Send(v)
}

// BroadcastDeviceInfo re-advertises our capabilities, called after a settings
// change so peers stop sending things the user just switched off.
func (m *Manager) BroadcastDeviceInfo() {
	if m.LocalDeviceInfo == nil {
		return
	}
	m.Broadcast("", m.LocalDeviceInfo())
}

// ------------------------------------------- SessionHandler (forwarding)

func (m *Manager) OnMessage(s *Session, msgType protocol.MessageType, raw []byte) {
	if m.Handler != nil {
		m.Handler.OnMessage(s, msgType, raw)
	}
}

func (m *Manager) OnDeviceInfo(s *Session, info protocol.DeviceInfo) {
	if m.Handler != nil {
		m.Handler.OnDeviceInfo(s, info)
	}
}

func (m *Manager) OnUnpair(s *Session, msg protocol.Unpair) {
	if m.Handler != nil {
		m.Handler.OnUnpair(s, msg)
	}
}

func (m *Manager) OnClosed(s *Session, err error) {
	deviceID := s.DeviceID()

	m.mu.Lock()
	// Only forget the session if it is still the registered one; a replaced
	// duplicate closing later must not evict the connection that superseded it.
	if e, ok := m.sessions[deviceID]; ok && e.session == s {
		delete(m.sessions, deviceID)
	}
	m.mu.Unlock()

	if m.OnSessionChange != nil {
		m.OnSessionChange(deviceID, false)
	}
	if m.Handler != nil {
		m.Handler.OnClosed(s, err)
	}
}

File: mobile/lib/core/transport/connection_manager.dart
import 'dart:async';
import 'dart:io';
import 'dart:math';

import '../discovery/discovery_service.dart';
import '../protocol/messages.dart';
import 'framing.dart';
import 'handshake.dart';
import 'session.dart';

const Duration _dialTimeout = Duration(seconds: 5);
const Duration _reconnectEvery = Duration(seconds: 4);
const Duration _minBackoff = Duration(seconds: 2);
const Duration _maxBackoff = Duration(seconds: 60);

/// A device asking to join this ecosystem, already authenticated.
class PairingRequest {
  final String deviceId;
  final String name;
  final String platform;
  final String formFactor;
  final String publicKey;
  final String verificationCode;
  final String address;

  const PairingRequest({
    required this.deviceId,
    required this.name,
    required this.platform,
    required this.formFactor,
    required this.publicKey,
    required this.verificationCode,
    required this.address,
  });
}

/// The user's verdict on a [PairingRequest].
class PairingDecision {
  final bool accepted;
  final String reason;
  const PairingDecision(this.accepted, [this.reason = '']);
}

class _Backoff {
  DateTime next = DateTime.fromMillisecondsSinceEpoch(0);
  Duration duration = _minBackoff;
}

class _SessionEntry {
  final Session session;
  final bool outbound;
  const _SessionEntry(this.session, this.outbound);
}

/// Owns the TCP listener, keeps a session open to every trusted peer that is
/// online, and routes inbound connections by intent.
class ConnectionManager {
  final LocalInfo local;
  final DiscoveryService discovery;
  final PeerAuthorizer auth;
  final SessionHandler handler;

  /// Called when an untrusted device asks to join. May await the user.
  final Future<PairingDecision> Function(PairingRequest request)? onPairingRequest;

  /// Called for an inbound transfer connection. The callee owns the connection
  /// and must close it.
  final Future<void> Function(SecureConnection conn, DeviceInfo peer, TransferOffer offer)?
      onTransferOffer;

  /// Fires whenever a peer connects or disconnects.
  final void Function(String deviceId, bool connected)? onSessionChange;

  /// Supplies the device info sent when a session opens.
  final DeviceInfo Function() localDeviceInfo;

  ServerSocket? _server;
  Timer? _maintainTimer;

  final Map<String, _SessionEntry> _sessions = {};
  final Set<String> _dialing = {};
  final Map<String, _Backoff> _backoff = {};
  final _random = Random();

  ConnectionManager({
    required this.local,
    required this.discovery,
    required this.auth,
    required this.handler,
    required this.localDeviceInfo,
    this.onPairingRequest,
    this.onTransferOffer,
    this.onSessionChange,
  });

  int get port => _server?.port ?? 0;

  List<String> get connectedDevices => _sessions.keys.toList();

  bool isConnected(String deviceId) => _sessions.containsKey(deviceId);

  Session? session(String deviceId) => _sessions[deviceId]?.session;

  /// Binds the listener and begins maintaining sessions, returning the port
  /// actually bound. Announcing a port the listener failed to claim is what
  /// made peers report endless connection failures with no visible cause.
  Future<int> start() async {
    if (_server != null) return _server!.port;

    ServerSocket server;
    try {
      server = await ServerSocket.bind(InternetAddress.anyIPv4, transportPort, shared: false);
    } catch (_) {
      // Fall back to an ephemeral port rather than disabling the app entirely.
      server = await ServerSocket.bind(InternetAddress.anyIPv4, 0, shared: false);
    }

    _server = server;
    server.listen(
      _handleInbound,
      onError: (_) {},
      cancelOnError: false,
    );

    _maintainTimer = Timer.periodic(_reconnectEvery, (_) => _connectMissing());
    return server.port;
  }

  Future<void> stop() async {
    _maintainTimer?.cancel();
    _maintainTimer = null;

    final sessions = _sessions.values.toList();
    _sessions.clear();
    for (final entry in sessions) {
      await entry.session.close();
    }

    await _server?.close();
    _server = null;
  }

  // ------------------------------------------------------------- inbound

  Future<void> _handleInbound(Socket socket) async {
    HandshakeResult result;
    try {
      result = await acceptHandshake(socket: socket, local: local, auth: auth);
    } catch (_) {
      // acceptHandshake already wrote a reason and destroyed the socket.
      return;
    }

    switch (result.intent) {
      case Intent.pair:
        await _handlePairing(result, socket.remoteAddress.address);
        break;
      case Intent.session:
        await _adoptSession(result, outbound: false);
        break;
      case Intent.transfer:
        await _handleInboundTransfer(result);
        break;
      default:
        await result.connection.close();
    }
  }

  Future<void> _handlePairing(HandshakeResult result, String address) async {
    try {
      if (onPairingRequest == null) {
        await result.connection.writeJson({
          'type': MsgType.pairingResp,
          'device_id': local.identity.deviceId,
          'accepted': false,
          'reason': 'this device cannot accept pairing right now',
        });
        return;
      }

      final decision = await onPairingRequest!(PairingRequest(
        deviceId: result.peer.deviceId,
        name: result.peer.name,
        platform: result.peer.platform,
        formFactor: result.peer.formFactor,
        publicKey: result.peerPublicKey,
        verificationCode: result.verificationCode,
        address: address,
      ));

      await result.connection.writeJson({
        'type': MsgType.pairingResp,
        'device_id': local.identity.deviceId,
        'name': local.name,
        'accepted': decision.accepted,
        if (decision.reason.isNotEmpty) 'reason': decision.reason,
      });

      if (decision.accepted) {
        // Connect straight away rather than waiting for the next tick, so the
        // device turns green the moment the user taps Accept.
        reconnectNow(result.peer.deviceId);
      }
    } catch (_) {
      // The peer hung up mid-decision; nothing left to do.
    } finally {
      await result.connection.close();
    }
  }

  Future<void> _handleInboundTransfer(HandshakeResult result) async {
    if (onTransferOffer == null) {
      await result.connection.close();
      return;
    }

    try {
      final message = await result.connection.readJson().timeout(const Duration(seconds: 30));
      if (message['type'] != MsgType.transferOffer) {
        await result.connection.close();
        return;
      }
      // The callee owns the connection from here, including closing it.
      await onTransferOffer!(
        result.connection,
        result.peer,
        TransferOffer.fromJson(message),
      );
    } catch (_) {
      await result.connection.close();
    }
  }

  // ------------------------------------------------------------ outbound

  Future<void> _connectMissing() async {
    for (final peer in discovery.peers) {
      final key = auth.trustedKey(peer.deviceId);
      if (key == null) continue;
      if (!_shouldDial(peer.deviceId)) continue;

      _dialing.add(peer.deviceId);
      // Deliberately not awaited: one slow peer must never delay the rest.
      unawaited(_dialPeer(peer, key));
    }
  }

  bool _shouldDial(String deviceId) {
    if (_sessions.containsKey(deviceId)) return false;
    if (_dialing.contains(deviceId)) return false;

    final backoff = _backoff[deviceId];
    if (backoff != null && DateTime.now().isBefore(backoff.next)) return false;
    return true;
  }

  Future<void> _dialPeer(DiscoveryMessage peer, String expectedKey) async {
    try {
      final result = await dialHandshake(
        host: peer.ip,
        port: peer.tcpPort,
        local: local,
        intent: Intent.session,
        expectedKey: expectedKey,
        timeout: _dialTimeout,
      );
      _backoff.remove(peer.deviceId);
      await _adoptSession(result, outbound: true);
    } catch (_) {
      _noteFailure(peer.deviceId);
    } finally {
      _dialing.remove(peer.deviceId);
    }
  }

  void _noteFailure(String deviceId) {
    final backoff = _backoff.putIfAbsent(deviceId, _Backoff.new);
    backoff.duration = Duration(
      milliseconds: min(backoff.duration.inMilliseconds * 2, _maxBackoff.inMilliseconds),
    );
    // Jitter keeps a roomful of devices from retrying in lockstep after the
    // access point restarts.
    final jitter = _random.nextInt(max(1, backoff.duration.inMilliseconds ~/ 4));
    backoff.next = DateTime.now().add(backoff.duration + Duration(milliseconds: jitter));
  }

  /// Clears any backoff for a device and tries it immediately.
  void reconnectNow(String deviceId) {
    _backoff.remove(deviceId);
    discovery.announce();

    final peer = discovery.peer(deviceId);
    final key = auth.trustedKey(deviceId);
    if (peer == null || key == null || !_shouldDial(deviceId)) return;

    _dialing.add(deviceId);
    unawaited(_dialPeer(peer, key));
  }

  /// Retries every trusted, currently-visible peer at once, clearing backoff.
  ///
  /// Called when the app returns to the foreground: a stretch backgrounded may
  /// have let Android throttle timers under Doze, so peers can look stale even
  /// though they never actually left the network. This collapses the wait back
  /// to "now" instead of the remaining backoff or the next maintenance tick.
  void retryAllNow() {
    discovery.announce();
    for (final peer in discovery.peers) {
      if (auth.trustedKey(peer.deviceId) == null) continue;
      _backoff.remove(peer.deviceId);
      if (!_shouldDial(peer.deviceId)) continue;
      _dialing.add(peer.deviceId);
      unawaited(_dialPeer(peer, auth.trustedKey(peer.deviceId)!));
    }
  }

  // ------------------------------------------------------------ sessions

  /// Installs a new session, resolving the case where both devices dialled each
  /// other at the same moment.
  ///
  /// Both ends apply the same rule — keep the connection dialled by whichever
  /// device has the smaller ID — so they independently agree on which socket
  /// survives, instead of each closing the other's and ending up with none.
  Future<void> _adoptSession(HandshakeResult result, {required bool outbound}) async {
    final deviceId = result.peer.deviceId;
    final existing = _sessions[deviceId];

    if (existing != null) {
      final preferOutbound = local.identity.deviceId.compareTo(deviceId) < 0;
      if (existing.outbound == preferOutbound || outbound != preferOutbound) {
        await result.connection.close();
        return;
      }
      _sessions.remove(deviceId);
      await existing.session.close();
    }

    final session = Session(
      connection: result.connection,
      peerInfo: result.peer,
      handler: SessionHandler(
        onMessage: handler.onMessage,
        onDeviceInfo: handler.onDeviceInfo,
        onUnpair: handler.onUnpair,
        onClosed: (s, error) {
          // Only forget the session if it is still the registered one; a
          // replaced duplicate closing later must not evict its successor.
          if (identical(_sessions[s.deviceId]?.session, s)) {
            _sessions.remove(s.deviceId);
            onSessionChange?.call(s.deviceId, false);
          }
          handler.onClosed?.call(s, error);
        },
      ),
    );

    _sessions[deviceId] = _SessionEntry(session, outbound);
    onSessionChange?.call(deviceId, true);

    unawaited(session.send(localDeviceInfo().toJson()).catchError((_) {}));
    unawaited(session.run());
  }

  Future<void> disconnect(String deviceId) async {
    final entry = _sessions.remove(deviceId);
    await entry?.session.close();
  }

  /// Sends a message to every connected peer that advertised [capability].
  /// An empty capability sends to everyone.
  Future<void> broadcast(String capability, Map<String, dynamic> message) async {
    for (final entry in _sessions.values.toList()) {
      if (capability.isNotEmpty && !entry.session.supports(capability)) continue;
      try {
        await entry.session.send(message);
      } catch (_) {
        await entry.session.close();
      }
    }
  }

  Future<void> sendTo(String deviceId, Map<String, dynamic> message) async {
    final entry = _sessions[deviceId];
    if (entry == null) throw StateError('that device is not connected');
    await entry.session.send(message);
  }

  /// Re-advertises capabilities after a settings change, so peers stop sending
  /// things the user just switched off.
  Future<void> broadcastDeviceInfo() => broadcast('', localDeviceInfo().toJson());
}
